File: mobile-rn/src/screens/AccountScreen.js
import { useEffect, useState } from "react";
import { Alert, ScrollView, StyleSheet, Text, View } from "react-native";
import { colors } from "../theme";
import { Field, PrimaryButton, SecondaryButton, DangerButton } from "../components/ui";
import { api, ApiError } from "../api";
import { useAuth } from "../context/AuthContext";

export default function AccountScreen({ navigation }) {
  const { auth, setAuth, patchAuth } = useAuth();

  const [storeName, setStoreName] = useState(auth?.storeName || "");
  const [profileMsg, setProfileMsg] = useState({ text: "", error: false });
  const [profileSubmitting, setProfileSubmitting] = useState(false);

  const [currentPin, setCurrentPin] = useState("");
  const [newPin, setNewPin] = useState("");
  const [pinMsg, setPinMsg] = useState({ text: "", error: false });
  const [pinSubmitting, setPinSubmitting] = useState(false);

  const [deleting, setDeleting] = useState(false);

  useEffect(() => {
    if (!auth) return;
    api
      .me(auth.token)
      .then((data) => {
        setStoreName(data.storeName || "");
        patchAuth({ storeName: data.storeName || "" });
      })
      .catch((err) => {
        if (err instanceof ApiError && err.status === 401) {
          setAuth(null);
          navigation.goBack();
        }
      });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  if (!auth) return null;

  async function handleProfileSubmit() {
    setProfileMsg({ text: "", error: false });
    setProfileSubmitting(true);
    try {
      const data = await api.updateProfile(auth.token, { storeName: storeName.trim() });
      patchAuth({ storeName: data.storeName || "" });
      setProfileMsg({ text: "Profil mis à jour.", error: false });
    } catch (err) {
      if (err instanceof ApiError && err.status === 401) {
        setAuth(null);
        navigation.goBack();
        return;
      }
      setProfileMsg({ text: err.message, error: true });
    } finally {
      setProfileSubmitting(false);
    }
  }

  async function handlePinSubmit() {
    setPinMsg({ text: "", error: false });
    setPinSubmitting(true);
    try {
      await api.updateProfile(auth.token, { currentPin: currentPin.trim(), newPin: newPin.trim() });
      setCurrentPin("");
      setNewPin("");
      setPinMsg({ text: "Code PIN mis à jour.", error: false });
    } catch (err) {
      if (err instanceof ApiError && err.status === 401) {
        setAuth(null);
        navigation.goBack();
        return;
      }
      setPinMsg({ text: err.message, error: true });
    } finally {
      setPinSubmitting(false);
    }
  }

  function handleLogout() {
    if (auth) api.logout(auth.token).catch(() => {});
    setAuth(null);
    navigation.goBack();
  }

  function confirmDelete() {
    Alert.alert(
      "Supprimer le compte",
      "Supprimer définitivement votre compte et toutes vos annonces ? Cette action est irréversible.",
      [
        { text: "Annuler", style: "cancel" },
        { text: "Supprimer", style: "destructive", onPress: handleDelete }
      ]
    );
  }

  async function handleDelete() {
    setDeleting(true);
    try {
      await api.deleteAccount(auth.token);
      setAuth(null);
      navigation.goBack();
    } catch (err) {
      Alert.alert("Erreur", err instanceof ApiError ? err.message : "Une erreur s'est produite.");
    } finally {
      setDeleting(false);
    }
  }

  return (
    <ScrollView contentContainerStyle={styles.container}>
      <Text style={styles.phone}>Numéro : {auth.maskedPhone}</Text>

      <Text style={styles.label}>Nom de la boutique</Text>
      <Field placeholder="ex. Boutique Fatou" maxLength={40} value={storeName} onChangeText={setStoreName} style={styles.gap} />
      <PrimaryButton title="Enregistrer le profil" onPress={handleProfileSubmit} loading={profileSubmitting} />
      {!!profileMsg.text && <Text style={[styles.msg, profileMsg.error && styles.msgError]}>{profileMsg.text}</Text>}

      <View style={styles.divider} />

      <Text style={styles.label}>Changer le code PIN</Text>
      <Field placeholder="Code PIN actuel" secureTextEntry keyboardType="number-pad" maxLength={6} value={currentPin} onChangeText={setCurrentPin} style={styles.gap} />
      <Field
        placeholder="Nouveau code PIN (4 à 6 chiffres)"
        secureTextEntry
        keyboardType="number-pad"
        maxLength={6}
        value={newPin}
        onChangeText={setNewPin}
        style={styles.gap}
      />
      <PrimaryButton title="Changer le code PIN" onPress={handlePinSubmit} loading={pinSubmitting} />
      {!!pinMsg.text && <Text style={[styles.msg, pinMsg.error && styles.msgError]}>{pinMsg.text}</Text>}

      <View style={styles.divider} />

      <View style={styles.dangerRow}>
        <SecondaryButton title="Se déconnecter" onPress={handleLogout} style={styles.flex1} />
        <DangerButton title={deleting ? "Suppression…" : "Supprimer mon compte"} onPress={confirmDelete} style={styles.flex1} />
      </View>
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: { padding: 22 },
  phone: { color: colors.textDim, fontSize: 11.5, marginBottom: 18 },
  label: { color: colors.teal, fontSize: 10.5, marginBottom: 6 },
  gap: { marginBottom: 10 },
  msg: { color: colors.textDim, fontSize: 11, marginTop: 6 },
  msgError: { color: colors.error },
  divider: { height: 1, backgroundColor: colors.hairline, marginVertical: 20 },
  dangerRow: { flexDirection: "row", gap: 8 },
  flex1: { flex: 1 }
});

File: mobile-flutter/lib/screens/account_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../services/api.dart';
import '../services/auth_provider.dart';
import '../theme/colors.dart';

class AccountScreen extends StatefulWidget {
  const AccountScreen({super.key});

  @override
  State<AccountScreen> createState() => _AccountScreenState();
}

class _AccountScreenState extends State<AccountScreen> {
  final _storeNameCtrl = TextEditingController();
  final _currentPinCtrl = TextEditingController();
  final _newPinCtrl = TextEditingController();

  String _profileMsg = '';
  bool _profileError = false;
  bool _profileSubmitting = false;

  String _pinMsg = '';
  bool _pinError = false;
  bool _pinSubmitting = false;

  bool _deleting = false;

  @override
  void initState() {
    super.initState();
    final auth = context.read<AuthProvider>().auth;
    _storeNameCtrl.text = auth?.storeName ?? '';
    if (auth != null) {
      Api.instance.me(auth.token).then((data) {
        if (!mounted) return;
        _storeNameCtrl.text = (data['storeName'] as String?) ?? '';
        context.read<AuthProvider>().patchStoreName((data['storeName'] as String?) ?? '');
      }).catchError((err) {
        if (err is ApiException && err.status == 401 && mounted) {
          context.read<AuthProvider>().setAuth(null);
          Navigator.of(context).pop();
        }
      });
    }
  }

  Future<void> _saveProfile() async {
    final auth = context.read<AuthProvider>().auth;
    if (auth == null) return;
    setState(() {
      _profileMsg = '';
      _profileSubmitting = true;
    });
    try {
      final data = await Api.instance.updateProfile(auth.token, {'storeName': _storeNameCtrl.text.trim()});
      if (!mounted) return;
      await context.read<AuthProvider>().patchStoreName((data['storeName'] as String?) ?? '');
      setState(() {
        _profileMsg = 'Profil mis à jour.';
        _profileError = false;
      });
    } catch (err) {
      if (err is ApiException && err.status == 401) {
        await context.read<AuthProvider>().setAuth(null);
        if (mounted) Navigator.of(context).pop();
        return;
      }
      setState(() {
        _profileMsg = err is ApiException ? err.message : 'Une erreur s\'est produite.';
        _profileError = true;
      });
    } finally {
      if (mounted) setState(() => _profileSubmitting = false);
    }
  }

  Future<void> _changePin() async {
    final auth = context.read<AuthProvider>().auth;
    if (auth == null) return;
    setState(() {
      _pinMsg = '';
      _pinSubmitting = true;
    });
    try {
      await Api.instance.updateProfile(auth.token, {
        'currentPin': _currentPinCtrl.text.trim(),
        'newPin': _newPinCtrl.text.trim(),
      });
      _currentPinCtrl.clear();
      _newPinCtrl.clear();
      setState(() {
        _pinMsg = 'Code PIN mis à jour.';
        _pinError = false;
      });
    } catch (err) {
      if (err is ApiException && err.status == 401) {
        await context.read<AuthProvider>().setAuth(null);
        if (mounted) Navigator.of(context).pop();
        return;
      }
      setState(() {
        _pinMsg = err is ApiException ? err.message : 'Une erreur s\'est produite.';
        _pinError = true;
      });
    } finally {
      if (mounted) setState(() => _pinSubmitting = false);
    }
  }

  Future<void> _logout() async {
    final auth = context.read<AuthProvider>().auth;
    if (auth != null) Api.instance.logout(auth.token).catchError((_) => null);
    await context.read<AuthProvider>().setAuth(null);
    if (mounted) Navigator.of(context).pop();
  }

  Future<void> _confirmDelete() async {
    final confirmed = await showDialog<bool>(
      context: context,
      builder: (ctx) => AlertDialog(
        backgroundColor: SnapyColors.panel,
        title: const Text('Supprimer le compte', style: TextStyle(color: SnapyColors.paper)),
        content: const Text(
          'Supprimer définitivement votre compte et toutes vos annonces ? Cette action est irréversible.',
          style: TextStyle(color: SnapyColors.textDim),
        ),
        actions: [
          TextButton(onPressed: () => Navigator.pop(ctx, false), child: const Text('Annuler')),
          TextButton(onPressed: () => Navigator.pop(ctx, true), child: const Text('Supprimer', style: TextStyle(color: SnapyColors.error))),
        ],
      ),
    );
    if (confirmed != true) return;

    final auth = context.read<AuthProvider>().auth;
    if (auth == null) return;
    setState(() => _deleting = true);
    try {
      await Api.instance.deleteAccount(auth.token);
      await context.read<AuthProvider>().setAuth(null);
      if (mounted) Navigator.of(context).pop();
    } catch (err) {
      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text(err is ApiException ? err.message : 'Une erreur s\'est produite.')),
        );
      }
    } finally {
      if (mounted) setState(() => _deleting = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final auth = context.watch<AuthProvider>().auth;
    if (auth == null) return const SizedBox.shrink();

    return Scaffold(
      appBar: AppBar(title: const Text('Mon compte')),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(22),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Text('Numéro : ${auth.maskedPhone}', style: const TextStyle(color: SnapyColors.textDim, fontSize: 11.5)),
            const SizedBox(height: 18),
            const Text('Nom de la boutique', style: TextStyle(color: SnapyColors.teal, fontSize: 10.5)),
            const SizedBox(height: 6),
            TextField(controller: _storeNameCtrl, maxLength: 40, decoration: const InputDecoration(hintText: 'ex. Boutique Fatou')),
            ElevatedButton(
              onPressed: _profileSubmitting ? null : _saveProfile,
              child: _profileSubmitting
                  ? const SizedBox(width: 20, height: 20, child: CircularProgressIndicator(strokeWidth: 2, color: SnapyColors.amberOn))
                  : const Text('Enregistrer le profil'),
            ),
            if (_profileMsg.isNotEmpty)
              Padding(
                padding: const EdgeInsets.only(top: 6),
                child: Text(_profileMsg, style: TextStyle(color: _profileError ? SnapyColors.error : SnapyColors.textDim, fontSize: 11)),
              ),
            const Divider(height: 40, color: SnapyColors.hairline),
            const Text('Changer le code PIN', style: TextStyle(color: SnapyColors.teal, fontSize: 10.5)),
            const SizedBox(height: 6),
            TextField(
              controller: _currentPinCtrl,
              obscureText: true,
              keyboardType: TextInputType.number,
              maxLength: 6,
              decoration: const InputDecoration(hintText: 'Code PIN actuel'),
            ),
            const SizedBox(height: 10),
            TextField(
              controller: _newPinCtrl,
              obscureText: true,
              keyboardType: TextInputType.number,
              maxLength: 6,
              decoration: const InputDecoration(hintText: 'Nouveau code PIN (4 à 6 chiffres)'),
            ),
            ElevatedButton(
              onPressed: _pinSubmitting ? null : _changePin,
              child: _pinSubmitting
                  ? const SizedBox(width: 20, height: 20, child: CircularProgressIndicator(strokeWidth: 2, color: SnapyColors.amberOn))
                  : const Text('Changer le code PIN'),
            ),
            if (_pinMsg.isNotEmpty)
              Padding(
                padding: const EdgeInsets.only(top: 6),
                child: Text(_pinMsg, style: TextStyle(color: _pinError ? SnapyColors.error : SnapyColors.textDim, fontSize: 11)),
              ),
            const Divider(height: 40, color: SnapyColors.hairline),
            Row(
              children: [
                Expanded(child: OutlinedButton(onPressed: _logout, child: const Text('Se déconnecter'))),
                const SizedBox(width: 8),
                Expanded(
                  child: OutlinedButton(
                    style: OutlinedButton.styleFrom(foregroundColor: SnapyColors.error, side: const BorderSide(color: Color(0x66E8756A))),
                    onPressed: _deleting ? null : _confirmDelete,
                    child: Text(_deleting ? 'Suppression…' : 'Supprimer mon compte'),
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
